src/validation.cpp

```diff
@@ -3832,6 +3832,18 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
     if (ppindex)
         *ppindex = pindex;
 
+    const auto msg = strprintf("saw new valid header hash=%s", hash.ToString());
```

Comment:
Maybe add an anti-footgun note here, since I almost made that mistake:

```
// Not all valid headers have a known height, so we don't log pindex->nHeight.
```

Reply:
Also, let's capitalize the log message: `Saw new …`

Reply:
How do you have a valid header without a known height? If you don't know all the ancestors, you can't tell if the `nBits` value is valid, which we check in `ContextualCheckBlockHeader` earlier in this function... (A missing prev block should result in `return Invalid(BLOCK_MISSING_PREV)` prior to that) The only case where we don't do all that is if we're looking at the genesis block header, but we know its height is 0... What am I missing here?

Reply:
`AddToBlockIndex` checks if `hashPrevBlock` is already in the index, and if not it won't set `pprev` and `nHeight` on the `CBlockIndex` entry it returns. But I didn't check if that condition can ever be reached from the call in `AcceptBlockHeader`.

Reply:
The only other call site for `AddToBlockIndex()` is `LoadGenesisBlock()`, so I guess that's the only reason for the (generic looking) check of `hashPrevBlock`.

Reply:
Indeed this code is pretty ancient (and could use a comment): https://github.com/bitcoin/bitcoin/blame/f3ae51dcced8a16175426051ce888130cc2493af/src/main.cpp#L2056-L2063